### DISCUSSION TITLE
Playwright integration testing

### DISCUSSION THREAD

**Discussion by sand4rt** [nextauthjs/next-auth#12183] [Help]
### What is the improvement or update you wish to see?

Documentation on how to use playwright + next-auth

### Is there any context that might help us understand?

I am using Playwright with the Next App Router and Mock Service Worker (MSW) to mock all API calls to a rest backend and i would like to mock the authenticated user as well.

I made an snapshot of the auth state with `await page.context().storageState({ path: authFile });`: https://playwright.dev/docs/auth#basic-shared-account-in-all-tests, However, the generated authFile appears to be insufficient:

```json
{
  "cookies": [
    {
      "name": "next-auth.session-token",
      "value": "MY_CUSTOM_JWT_WITH_A_REALLY_LARGE_EXPIRE_DATE",
      "domain": "localhost",
      "path": "/",
      "expires": -1,
      "httpOnly": true,
      "secure": false,
      "sameSite": "Lax"
    }
  ],
  "origins": []
}
```

I’ve noticed that the Microsoft authentication provider performs additional API calls to verify the validity of the token. I attempted to mock one of these calls using MSW:

```ts
// This mock intercepts API calls in the next backend only
import { createServer } from '@mswjs/http-middleware';
const httpServer = createServer(...[
    http.get('https://login.microsoftonline.com/*', () => HttpResponse.json({
        claim_types_supported: ['normal'],
        claims_parameter_supported: false,
        grant_types_supported: ['authorization_code', 'implicit'],
        request_parameter_supported: false,
        request_uri_parameter_supported: false,
        require_request_uri_registration: false,
        //...
    })),
]);
```

But i guess there is more to it?

### Does the docs page already exist? Please link to it.

https://next-auth.js.org/guides/testing

**Comment by MILLERMARRU** [comment]
The key insight: **don't try to mock the Microsoft OAuth flow** — OIDC involves a discovery document fetch, JWKS endpoint, token exchange, and `userinfo` calls. Intercepting all of them with MSW is fragile and couples your tests to OAuth internals your app doesn't care about.

Instead, bypass OAuth entirely and inject a valid session cookie directly. NextAuth/Auth.js only cares about the signed JWT in the cookie — if it's valid (correct secret, not expired), `useSession()`, `auth()`, and `getToken()` all work as normal.

## Step 1 — Create a session cookie helper

```ts
// tests/helpers/auth.ts
import { SignJWT } from 'jose';

interface TestUser {
  id: string;
  email: string;
  name: string;
}

export async function buildSessionCookie(user: TestUser = {
  id: 'test-user-id',
  email: 'test@example.com',
  name: 'Test User',
}) {
  const secret = new TextEncoder().encode(
    process.env.AUTH_SECRET ?? process.env.NEXTAUTH_SECRET
  );

  const token = await new SignJWT({
    sub: user.id,
    email: user.email,
    name: user.name,
    picture: null,
    jti: crypto.randomUUID(),
  })
    .setProtectedHeader({ alg: 'HS256' })
    .setIssuedAt()
    .setExpirationTime('1h')
    .sign(secret);

  return {
    // Auth.js v5: 'authjs.session-token'
    // NextAuth v4: 'next-auth.session-token'
    name: 'authjs.session-token',
    value: token,
    domain: 'localhost',
    path: '/',
    httpOnly: true,
    secure: false,
    sameSite: 'Lax' as const,
  };
}
```

## Step 2 — Create a reusable Playwright fixture

```ts
// tests/fixtures.ts
import { test as base, type Page } from '@playwright/test';
import { buildSessionCookie } from './helpers/auth';

type Fixtures = {
  authenticatedPage: Page;
};

export const test = base.extend<Fixtures>({
  authenticatedPage: async ({ page }, use) => {
    // Must visit a page first so the browser has a domain context
    await page.goto('/');
    await page.context().addCookies([await buildSessionCookie()]);
    await use(page);
  },
});

export { expect } from '@playwright/test';
```

## Step 3 — Use in tests

```ts
// tests/dashboard.spec.ts
import { test, expect } from './fixtures';

test('shows protected content when authenticated', async ({ authenticatedPage }) => {
  await authenticatedPage.goto('/dashboard');
  await expect(authenticatedPage.getByText('Welcome')).toBeVisible();
});

test('can test with a specific user', async ({ page }) => {
  await page.goto('/');
  await page.context().addCookies([
    await buildSessionCookie({ id: 'admin-id', email: 'admin@example.com', name: 'Admin' }),
  ]);
  await page.goto('/admin');
  await expect(page.getByRole('heading', { name: 'Admin Panel' })).toBeVisible();
});
```

## Why not MSW for Microsoft OAuth?

Microsoft's OIDC flow involves at minimum:
1. `/.well-known/openid-configuration` — discovery document
2. `/discovery/v2.0/keys` — JWKS for token verification
3. `/oauth2/v2.0/token` — token exchange
4. `/oidc/userinfo` — user profile

Mocking all four in MSW + making sure they satisfy the `openid-client` validation logic is a significant maintenance burden. The session cookie approach above is stable across provider changes, doesn't require network mocking, and runs faster.

Make sure `AUTH_SECRET` (or `NEXTAUTH_SECRET`) in your `.env.test` matches the one your dev server uses — that's the only requirement.
